controllers/agents/svc/serviceclass_controller.go

```diff
@@ -0,0 +1,240 @@
+
+	if err != nil || services == nil {
+		reconcileLog.Error(err, "Failed to retrieve resources", "gvr", mapping.Resource)
+		return ctrl.Result{}, nil
+	}
+
+	requeue := false
+	err = r.CreateRegisteredServices(ctx, &serviceClass, *services)
+	if err != nil {
+		reconcileLog.Error(err, "Failed to write registered services", "namespace", req.Namespace, "name", req.Name)
+		requeue = true
+	}
+	err = r.Client.Status().Update(ctx, &serviceClass)
+	if err != nil {
+		requeue = true
+	}
+	return ctrl.Result{Requeue: requeue}, nil
+}
+
+func (r *ServiceClassReconciler) CreateRegisteredServices(ctx context.Context, serviceClass *v1alpha1.ServiceClass, services unstructured.UnstructuredList) error {
+	l := log.FromContext(ctx)
+	mappings := map[string]*jsonpath.JSONPath{}
+	for _, mapping := range serviceClass.Spec.Resource.ServiceEndpointDefinitionMapping {
+		path := jsonpath.New("")
+		err := path.Parse(fmt.Sprintf("{%s}", mapping.JsonPath))
+		if err != nil {
+			return err
+		}
+		mappings[mapping.Name] = path
+	}
+
+	config, remote_namespace, err := r.getPrimazaKubeconfig(ctx, serviceClass.Namespace)
+	if err != nil {
+		return err
+	}
+	l.Info("remote cluster", "address", config.Host)
+
+	status := workercluster.TestConnection(ctx, config)
+	serviceClass.Status.Conditions = append(serviceClass.Status.Conditions, metav1.Condition{
+		Type:    "Connection",
+		Message: status.Message,
+		Reason:  string(status.Reason),
+		Status:  metav1.ConditionStatus(status.State),
+	})
+	if status.State == v1alpha1.ClusterEnvironmentStateOffline {
+		return fmt.Errorf("Failed to connect to cluster")
+	}
+
+	remote_client, err := client.New(config, client.Options{
+		Scheme: r.RemoteScheme,
+		Mapper: r.Mapper,
+	})
+	if err != nil {
+		return err
+	}
+
+	for _, data := range services.Items {
+		sedMappings, err := LookupServiceEndpointDescriptor(mappings, data)
+		if err != nil {
+			l.Error(err, "Failed to lookup service endpoint descriptor values",
+				"name", data.GetName(),
+				"namespace", data.GetNamespace(),
+				"gvk", data.GroupVersionKind())
+		}
+
+		rs := v1alpha1.RegisteredService{
+			ObjectMeta: metav1.ObjectMeta{
+				// FIXME(sadlerap): this could cause naming conflicts; we need
+				// to take into account the type of resource somehow.
+				Name:      data.GetName(),
+				Namespace: remote_namespace,
+			},
+			Spec: v1alpha1.RegisteredServiceSpec{
+				ServiceEndpointDefinition: sedMappings,
+				ServiceClassIdentity:      serviceClass.Spec.ServiceClassIdentity,
+				HealthCheck:               serviceClass.Spec.HealthCheck,
+			},
+		}
+
+		if serviceClass.Spec.Constraints != nil {
+			rs.Spec.Constraints = &v1alpha1.RegisteredServiceConstraints{
+				Environments: serviceClass.Spec.Constraints.Environments,
+			}
+		}
+
+		if err := remote_client.Create(ctx, &rs); err != nil {
+			l.Error(err, "Failed to create registered service",
+				"service", data.GetName(),
+				"namespace", remote_namespace)
+			return err
+		}
```

Comment:
`CreateOrUpdate` or fallback for now to `client.IgnoreAlreadyExisting`
Also collect all the errors and then join them together.

Reply:
fixed